core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -66,20 +66,36 @@ class BaseUpdatePartitionSpec implements UpdatePartitionSpec {
     this.caseSensitive = true;
     this.base = ops.current();
     this.formatVersion = base.formatVersion();
-    this.spec = base.spec();
+    fromSpec(base.spec());
+  }
+
+  @Override
+  public UpdatePartitionSpec fromSpec(PartitionSpec partitionSpec) {
+    // Clear all changes
```

Comment:
Implementation wise, I would rather we return a new object here? Couldn't we add a constructor

BaseUpdatePartitionSpec(TableOperations ops, PartitionSpec spec)

That way we could keep all of the variables above final?

Reply:
I'm also wondering if we may have issues here with re-adding a transform that already exists in another spec. I think currently that's handled by re-using the existing transform.

Example

```
Add Identity (x) 
Remove Identity (x)
Add Identity (x)
```

In the current code there would only ever be one Identity(x) transform. Would we be able to maintain that if we did

```
Add Identity(x)
From(Unpartitioned)
Add Identity(x)
```

The above code should be a noop right?

Reply:
> Implementation wise, I would rather we return a new object here? Couldn't we add a constructor
> 
> BaseUpdatePartitionSpec(TableOperations ops, PartitionSpec spec)
> 
> That way we could keep all of the variables above final?

Modified and fixed revApi

Reply:
> I'm also wondering if we may have issues here with re-adding a transform that already exists in another spec. I think currently that's handled by re-using the existing transform.
> 
> Example
> 
> ```
> Add Identity (x) 
> Remove Identity (x)
> Add Identity (x)
> ```
> 
> In the current code there would only ever be one Identity(x) transform. Would we be able to maintain that if we did
> 
> ```
> Add Identity(x)
> From(Unpartitioned)
> Add Identity(x)
> ```
> 
> The above code should be a noop right?

I've added more specs, I hope that would answer your question.

Calling Add Identity(X), and then Remove Identity(X) within the same UpdatePartitionSpec throws an exception:

```
Cannot delete newly added field: 1001: id: identity(1)
java.lang.IllegalArgumentException: Cannot delete newly added field: 1001: id: identity(1)
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
	at org.apache.iceberg.BaseUpdatePartitionSpec.removeField(BaseUpdatePartitionSpec.java:247)
```

Reply:
> Calling Add Identity(X), and then Remove Identity(X) within the same UpdatePartitionSpec throws an exception:
> 
> ```
> Cannot delete newly added field: 1001: id: identity(1)
> java.lang.IllegalArgumentException: Cannot delete newly added field: 1001: id: identity(1)
> 	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:220)
> 	at org.apache.iceberg.BaseUpdatePartitionSpec.removeField(BaseUpdatePartitionSpec.java:247)
> ```

Not in the same command.
Two commands

Command 1
  Add Identity(id)
  commit()
  
Command 2
  From unpartitioned
  Add Identity(id)
  commit() // This should fail I think? or be a noop?

Reply:
Those seem to be the specs that @shanielh  added, resulting in a noop.

Reply:
@RussellSpitzer any chance to look at this? I've added the specs you wanted in the last iteration. Thanks 🙏